File: internal/manager/agent/files_unix.go
//go:build !windows

package agent

import "os"

func replaceAtomic(from, to string) error { return os.Rename(from, to) }

func syncDirectory(path string) error {
	dir, err := os.Open(path)
	if err != nil {
		return err
	}
	defer dir.Close()
	return dir.Sync()
}

func restrictPrivate(path string, directory bool) error {
	if directory {
		return os.Chmod(path, 0o700)
	}
	return os.Chmod(path, 0o600)
}

func applyPrivateMode(path string, source os.FileMode) error {
	mode := source.Perm() & 0o700
	if mode&0o400 == 0 {
		mode |= 0o400
	}
	return os.Chmod(path, mode)
}

func applyTargetPermissions(path, _ string, mode os.FileMode) error {
	return os.Chmod(path, mode.Perm())
}

func privatePermissionsOK(_ string, directory bool, mode os.FileMode) bool {
	want := os.FileMode(0o600)
	if directory {
		want = 0o700
	}
	return mode.Perm() == want
}

File: desktop/src-tauri/src/manager_core/agent/files_unix.rs
use std::fs::{self, OpenOptions};
use std::io;
use std::os::unix::fs::PermissionsExt;
use std::path::Path;

pub fn replace_atomic(from: &Path, to: &Path) -> io::Result<()> {
    fs::rename(from, to)
}

pub fn sync_directory(path: &Path) -> io::Result<()> {
    let dir = OpenOptions::new().read(true).open(path)?;
    dir.sync_all()
}

pub fn restrict_private(path: &Path, directory: bool) -> io::Result<()> {
    let mode = if directory { 0o700 } else { 0o600 };
    fs::set_permissions(path, fs::Permissions::from_mode(mode))
}

pub fn apply_private_mode(path: &Path, source_mode: u32) -> io::Result<()> {
    let mut mode = source_mode & 0o700;
    if mode & 0o400 == 0 {
        mode |= 0o400;
    }
    fs::set_permissions(path, fs::Permissions::from_mode(mode))
}

pub fn apply_target_permissions(path: &Path, _target: &Path, mode: u32) -> io::Result<()> {
    fs::set_permissions(path, fs::Permissions::from_mode(mode & 0o777))
}

pub fn private_permissions_ok(_path: &Path, directory: bool, mode: u32) -> bool {
    let want = if directory { 0o700 } else { 0o600 };
    mode & 0o777 == want
}
